### ISSUE TITLE
New Architecture - Considerations/Discussion

### ISSUE THREAD

**Issue by 8BallBomBom** [react-native-webrtc/react-native-webrtc#1557]
**Slight update on this.**
With React Native 0.76 the New Architecture has become default.
Time to start making some progress 😅

----

With the release of React Native 0.74 there have been some significant improvements to the upcoming New Architecture.
As it stands things are still experimental but sooner or later the New Architecture will become default.
Now seems like a good time to start working on supporting the new stuff.
Luckily we can still keep some backwards compatibility.

Over the coming weeks I'll be pushing changes up over [here](https://github.com/react-native-webrtc/react-native-webrtc/tree/new-arch) on the `new-arch` branch.
Then after a while we can start testing and eventually get a release out.

Pinning this as a place to discuss anything related to the upcoming changes.

**Comment by saghul** [comment]
Last I checked new arch is basically turbo modules in C++ but we can't do that here.

So well the minimum compatibility thing to keep using the bridge anyway.

**Comment by 8BallBomBom** [comment]
Probably should have mentioned that a little better ahaha 🤔 
But yes we will be keeping to Java and Objective-C but implementing the TurboModule and Fabric changes.
Both changes fall under the New Architecture.

But to fully take advantage of everything we would indeed need to change over to C++ 😅 
Something i'm willing to eventually put some time into.

**Comment by santhoshvai** [comment]
you can also count me into this, we can move to using turbo modules and fabric with a fair amount of work

**Comment by 8BallBomBom** [comment]
Been a little delayed on this but will be looking to put the first parts up soon.
TurboModule + Fabric - Android and TS first 😉 

---

Currently dealing with some Catalyst/Bridgeless related issues.
Android support seems to be semi working, iOS still needs some work.
This will be a disruptive update as some things have been moved around and requirements have changed a little.
But what counts is the obvious, this should be a drop in update.
Soon we can start testing 🫨 

---

Will be working on this again shortly.
Currently waiting to see if some of the New Arch issues are fixed in the coming RN release.

**Comment by scssrs** [comment]
I'm duplicating here the message from #1611, in the hope that someone here might be aware of the problem and how to solve it.

--------
Hello! We are using react-native-webrtc in our android app and have tried to migrate to the new RN architecture. However, we have a problem because RTCView does not see tracks in the stream. Everything looks like MediaStrem’s addTrack method began to work asynchronously and at the moment when RTCView tries to get tracks from the stream, the tracks have not yet been added to the stream.

Here is a code example of how we roughly use RTCView and everything worked fine before switching to the new architecture

```javascript

export const RemoteVideoView = (props: Props) => {
  const {
    mirror,
    objectFit,
    remoteVideoTrack,
    zOrder,
    style,
    accessibilityLabel,
  } = props

  const mediaStream = useMemo(() => {
    const stream = new MediaStream()
    stream.addTrack(remoteVideoTrack)
    return stream
  }, [remoteVideoTrack])

  return <RTCView
    streamURL={mediaStream?.toURL()}
    style={style}
    mirror={mirror}
    zOrder={zOrder}
    objectFit={objectFit}
    accessibilityLabel={accessibilityLabel}
  />
}

```
But now it displays a black screen and we get a message in  the console "No video stream for react tag: " + streamURL

If add a setTimeout before render, the video displayed

```javascript
export const RemoteVideoView = (props: Props) => {
  const {
    mirror,
    objectFit,
    remoteVideoTrack,
    zOrder,
    style,
    accessibilityLabel,
  } = props

  const [render, setRender] = useState<boolean>(false)

  useEffect(() => {
    if (remoteVideoTrack) {
      setTimeout(() => {
        setRender(true)
      }, 500)
    } else {
      setRender(false)
    }
  }, [remoteVideoTrack])

  const mediaStream = useMemo(() => {
    const stream = new MediaStream()
    stream.addTrack(remoteVideoTrack)
    return stream as any
  }, [remoteVideoTrack])

  return render
    ? <RTCView
    streamURL={mediaStream?.toURL()}
    style={style}
    mirror={mirror}
    zOrder={zOrder}
    objectFit={objectFit}
    accessibilityLabel={accessibilityLabel}
  />
    : null
}
```

But I really don't like this solution. Could someone please recommend a better solution?

**Comment by 8BallBomBom** [comment]
@scssrs Some things had to be changed around a little to work properly with the new arch, luckily the issue you posted is one of the things that was found and fixed during early testing.

That being said there are other issues currently preventing streams from being rendered properly.
Best option at the moment would be to wait for a release that supports the new arch and disable the new arch support in rn.

**Comment by scssrs** [comment]
@8BallBomBom Thanks for your answer! Could you tell approximately when we can expect the release?

**Comment by 8BallBomBom** [comment]
That is hard to determine currently but hopefully before the end of the year.
Currently waiting to see how the new arch develops in the upcoming react native releases.
As it stands there are still some outstanding weird issues.

**Comment by NOT-HAL9000** [comment]
Would this include any new developments toward being able to have frame encryptors for e2ee?

**Comment by 8BallBomBom** [comment]
Sadly no, that may come at a later time, all depends on implementation.

**Comment by arunrphysicianhelper** [comment]
Hey team, wondering whats the progress on this? Does this block use of 0.76 react native version? 
I am getting an issue where Ice Connection is always failing. Wasn't sure if this works needs to be done to support this react native version. Thanks team.

**Comment by saghul** [comment]
With all the backwards compatibility they added in 0.77 it should work out of the box I think.

**Comment by 8BallBomBom** [comment]
That seems to be the case, been able to test and run without any issues since RN 0.76 and with newer versions backwards compatibility just improved bit by bit.

Will be looking to pickup where i left off at some point unless someone beats me to it, just been busy with a few other projects.

**Comment by mdrajibsk8** [comment]
I am using RN 0.76, and I'm using socket & WEBTRC for the calling. 

Issue - Not able to call with the different network while using the same network on both devices everything is working as expected. I tried a different version of WEBTRC also while using the different network in the log I'm getting the remoteStream.

For the STUN & TURN server - using multiple premium versions as well(Twilio, expressturn and created a STUN & TURN server using was as well)

Error log with different network - "ICE Connection State: failed - Connection failed. Check ICE candidates or TURN configuration."

**Comment by saghul** [comment]
That has nothing to do with new architecture. Please post in the forum.

**Comment by ponikar** [comment]
Hey folks, 
Just wanted to know I am able to run `"react-native-webrtc": "^124.0.4",` with new architecture enabled. How does it even possible if the library doesn't support new arch? 

My react native version is `"0.76.5"`
Am I missing something?

**Comment by 8BallBomBom** [comment]
We can close this as the module does seem to work with the latest RN due to the compatibility layer.
But it is still worth doing changes to move over to the new arch in some time.
Most likely performance improvements by doing so but a whole bunch of issues still.

**Comment by ponikar** [comment]
Thank you @saghul for the answer.
I agree with @8BallBomBom it will definitely going to worth migrating to new arch.

**Comment by nalladev** [comment]
Shouldn't we update here then? 

![Image](https://github.com/user-attachments/assets/5b9069ee-0146-4865-b087-b7cc6ca2cab8)

https://reactnative.directory

**Comment by saghul** [comment]
How does one update that?

**Comment by nalladev** [comment]
Probably in their repo

**Comment by nalladev** [comment]
https://github.com/react-native-community/directory

**Comment by saghul** [comment]
👍 I'll take a look!

**Comment by saghul** [comment]
Given that we need to still use the bridge (so no bridgeless), what is there missing to do?

**Comment by 8BallBomBom** [comment]
You would only need to remain using the old methods/bridge if running an older RN or deliberately running the old arch.
With the update that just went up comes full bridgeless/new arch support.
We could actually just strip out the old arch entirely but i figured keeping backwards compatibility would be a good thing.
Everyone is currently using the **Interop Compatibility Layer** which allows the module to stay alive under new RN releases.
Only issue we have now is time again, that layer will be removed in an update which is creeping up.

**Comment by saghul** [comment]
This plugin cannot work bridgeless. You need C++ for that, which implies a full rewrite, and there are many parts that are not implemented in C++ in WebRTC itself, like rendering or MediaCodec access on Android.

So my question is: what actual code changes are necessary here?

**Comment by 8BallBomBom** [comment]
I was under the impression that bridgeless works because of the codegen specs generating the C++ code when building.
Which can still be done and function with Java/Kotlin code for Android and Swift/Obj C for iOS.
If that isn't the case then we still fall into a big issue when the **Interop Compatibility Layer** is removed.
If i'm wrong then it would seem that i need to refocus my efforts and start on a C++ build 🤔 

It shows as bridgeless when running under the new arch but that isn't technically true as it uses a different bridge but that is from codegen creating the C++ glue between things, on the other hand it does mean everything is more direct.

**Comment by saghul** [comment]
> I was under the impression that bridgeless works because of the codegen specs generating the C++ code when building.
> Which can still be done and function with Java/Kotlin code for Android and Swift/Obj C for iOS.

Isn't that just for typing information? I haven't dug deep into the compat layer myself.



> If i'm wrong then it would seem that i need to refocus my efforts and start on a C++ build 🤔

I don't think _that_ is worth it. iOS would be easy because it's already ObjC++, but Androtid has many bits in pure Java, including rendering.

**Comment by 8BallBomBom** [comment]
At this point all i know is within the next few RN updates the compatibility layer is going to be removed entirely.
Which pretty much means this module will stop working unless converted to the new methods.
As a side project i'm going to see what it would take to do a C++ conversion, new arch only.

<details>
<summary>A general rundown of the new arch</summary>
<br />
Will React Native modules need to be written in C++ when the interop layer is removed?
<br />
No, React Native modules do not need to be written in C++ once the interop layer is removed.
<br />
While the New Architecture is built on a C++ foundation (JSI), you can continue to use platform-specific languages. 
<br /><br />
Here is how module development changes in the New Architecture:
<br /><br />

- **Platform-Specific Implementation:** You can still implement [Turbo Native Modules](https://reactnative.dev/docs/turbo-native-modules-introduction) using Kotlin/Java for Android and Swift/Objective-C for iOS.
- **Optional "Pure C++" Modules:** You now have the option to write [Pure C++ Turbo Modules](https://reactnative.dev/docs/the-new-architecture/pure-cxx-modules). These allow you to write your logic once in C++ and share it across all platforms (Android, iOS, Windows, macOS) without writing platform-specific code.
- **The Role of Codegen:** Instead of the interop layer, the New Architecture relies on [Codegen](https://reactnative.dev/docs/the-new-architecture/what-is-codegen). This tool takes a TypeScript or Flow specification and automatically generates the necessary C++ "glue code" for you. This allows your Java or Objective-C code to communicate with the JavaScript engine via JSI without you having to write manual C++ bindings.
- **Direct JSI Benefits:** By moving away from the old bridge, even your Kotlin or Swift modules benefit from [JSI (JavaScript Interface)](https://reactnative.dev/architecture/landing-page). This provides faster, synchronous communication and removes the overhead of JSON serialization. 

In summary, you will only need C++ if you want to create a cross-platform shared module or perform high-performance computations that benefit from a C++ implementation. For standard platform API access, Kotlin and Swift remain first-class citizens. 

</details>

**Comment by qusaieilouti99** [comment]
Hi guys @8BallBomBom  @saghul, sorry for pinging.
This library is the only library that brings WebRTC for react native, and many apps rely on it; soon, it will be the bottleneck that prevents users from migrating to the new architecture.

I suggest you provide a spec and introduction for other developers so they can help rewrite the library. They will start just forking, unless you have no plans to keep supporting the library in the future, so someone else can create another one
I'm not a native developer, but with a good spec and a good LLM, we can implement and verify each block against the specs without losing track; others maybe don't need LLM.

Currently, there is a huge room for enhancments its not just a "migration". 

Possible scenarios: 
1- We can just rewrite the shared logic in C++, the platform-specific via Kotlin/Swift or even Java/objc, all via a turbo module 
2- If we also want to have even faster bindings, an object-oriented approach, we can use Nitro modules. This opens a room for plugins to be added by the community later, like frame encryptors or their own Nitro View wrapper for the current "RTCVIEW", or even their own handshake like ZRTP. See this as an example of what is possible with Nitro: https://blog.margelo.com/whats-new-in-visioncamera-v5 
3- If you want to wait a bit more, there is NAPI support coming, maybe this year or early next year to Hermes, which will allow you to wrap an existing node native module, which I guess Chromium WebRTC is already one, and reuse it in react native, prebuilt, no JSI needed, no inconsistency between the different platforms.

I mean, I don't know your plans, but I just think if you don't have enough time for migration, and are still willing to keep the library, let's announce and bring other developers to help maintain and migrate. This is, IMHO, the power of OSS.

**Comment by qusaieilouti99** [comment]
Read this https://nitro.margelo.com/docs/resources/comparison
Helps with explaining the constraints and differences between turbo, nitro, expo, and legacy modules

**Comment by 8BallBomBom** [comment]
As it stands we're yet to figure out the exact direction but to get things started i did update the [new-arch](https://github.com/react-native-webrtc/react-native-webrtc/tree/new-arch) branch.
That includes backwards compatibility for RN as low as v0.76 along with full support for the new archtecture.
Whether we move forward with that, i'm not really sure.
Android is tested and working, iOS hasn't been tested as of yet.

For now I'm looking at nitro and also the possibility of a total rebuild entirely focused on C++ shared logic.

**Comment by saghul** [comment]
Looks like I missed a few messages here, so commenting now.

@8BallBomBom 

> At this point all i know is within the next few RN updates the compatibility layer is going to be removed entirely.

There were different layers, so we'd need to be sure if we are affected or not. It should be easy to test with a vanilla RN project using the latest released version.

> In summary, you will only need C++ if you want to create a cross-platform shared module or perform high-performance computations that benefit from a C++ implementation. For standard platform API access, Kotlin and Swift remain first-class citizens.

What I don't know is how that plays with the "bridgeless" feature.

@qusaieilouti99 

> This library is the only library that brings WebRTC for react native, and many apps rely on it; soon, it will be the bottleneck that prevents users from migrating to the new architecture.

Have you actually verified this yourself? What errors do you get?

> I mean, I don't know your plans, but I just think if you don't have enough time for migration, and are still willing to keep the library, let's announce and bring other developers to help maintain and migrate. This is, IMHO, the power of OSS.

After all this time with this issue open, I haven't seen concrete evidence of _what_ needs to happen. Right now it seems like the answer to that is _nothing_, because things Just Work (TM).

Moving to C++ would mean taking on a pretty sizable maintenance burden, even if we have LLMs to help.

**Comment by qusaieilouti99** [comment]
@saghul 
> What I don't know is how that plays with the "bridgeless" feature.

On releasing RN 82, they said https://reactnative.dev/blog/2025/10/08/react-native-0.82#interop-layers--3p-library-compatibility:

> We will keep the interop layers in the codebase for the foreseeable future. All the classes and functions that are required by the interop layers won’t be removed anytime soon. We will share further updates in the future regarding the removal of Interop Layers.
> 
> We’ve also verified that the 3P libraries that offer backward compatibility with both architectures will keep on working with 0.82, where New Architecture is the only architecture.



RN 84 https://reactnative.dev/blog/2026/02/11/react-native-0.84#legacy-architecture-components-removed: 

> In 0.83, we introduced the experimental RCT_REMOVE_LEGACY_ARCH flag to compile out Legacy Architecture code. In 0.84, this is now the default behaviour — Legacy Architecture code is no longer included in your iOS builds, reducing both build time and app size.
> 
> No breakages are expected for apps already on the New Architecture — the Interop Layer code required for compatibility remains in place

So, currently, let's say for a year, it will remain in place



@saghul 
> Have you actually verified this yourself? What errors do you get?

It's currently working for me because this layer keeps the bridge working.



@saghul 
> Moving to C++ would mean taking on a pretty sizable maintenance burden, even if we have LLMs to help.


This makes more sense to start planning from now, rather than having to do it after they remove the interop. You will need to migrate anyway soon, because react native team said that they plan to get rid of all the old arch leftovers incrementally, so it's not a decorative upgrade. Other libs measured actual performance boosts, better typesafety

**Comment by saghul** [comment]
Either of us is misunderstanding what they are saying.

They say they are removing legacy arch, that doesn't mean they are removing the interop layer.

**Comment by 8BallBomBom** [comment]
There isn't a set date for the interop compatibility layer to be removed but it is on the chopping block.
What i've implemented keeps a huge chunk of the current code and changes things over to a Turbo Module with Fabric.
That will at least survive the compatibility layer being removed and when that happens we just remove the backwards compatibility.
In the future maybe a shared C++ option would be nice but that would be a lot of work.
